### PR TITLE
compiler: Rework multi-level buffering

### PR DESCRIPTION
**Pull request by FabioLuporini**
See comments next to code for more info

### CODE REVIEW

File: devito/ir/support/syncs.py [hunk 1]
```diff
@@ -66,9 +66,16 @@ def __repr__(self):
 
     @property
     def imask(self):
-        ret = [self.handle.indices[d] if d.root in self.lock.locked_dimensions else FULL
-               for d in self.target.dimensions]
-        return IMask(*ret, getters=self.target.dimensions, function=self.function,
+        ret = []
+        for d in self.target.dimensions:
```
Comment (FabioLuporini) [inline devito/ir/support/syncs.py:70]: mere refactoring for code homogeneity (see below...)

File: devito/ir/support/syncs.py [hunk 2]
```diff
@@ -81,9 +88,19 @@ def __repr__(self):
 
     @property
     def imask(self):
-        ret = [(self.tindex, self.size) if d.root is self.dim.root else FULL
-               for d in self.target.dimensions]
-        return IMask(*ret, getters=self.target.dimensions, function=self.function,
+        ret = []
+        for d in self.target.dimensions:
+            if d.root is self.dim.root:
+                if self.target.is_regular:
```
Comment (FabioLuporini) [inline devito/ir/support/syncs.py:94]: added this; code path excited via PRO

File: devito/passes/clusters/buffering.py [hunk 3]
```diff
@@ -175,6 +176,12 @@ def callback(self, clusters, prefix, cache=None):
                 if b.size == 1 and not init_onread(b.function):
                     continue
 
+                # Special case: avoid initialization in the case of double
+                # (or multiple levels of) buffering because it will have been
+                # already performed
+                if b.size > 1 and b.multi_buffering:
```
Comment (FabioLuporini) [inline devito/passes/clusters/buffering.py:182]: code path excited via PRO

File: devito/passes/clusters/buffering.py [hunk 4]
```diff
@@ -347,64 +354,14 @@ def __init__(self, function, dim, d, accessv, cache, options, sregistry):
...
-                    v = ModuloDimension(name, xd, i, size)
-                    md = sregistry.setdefault('mds', (xd, i), v)
-                self.index_mapper[i] = md
-                self.sub_iterators[d.root].append(md)
+        # Initialize the buffer metadata. Depending on whether it's multi-level
+        # buffering (e.g., double buffering) or first-level, we need to perform
+        # different actions. Multi-level is trivial, because it essentially
+        # inherits metadata from the previous buffering level
+        if self.multi_buffering:
+            self.__init_multi_buffering__()
```
Comment (FabioLuporini) [inline devito/passes/clusters/buffering.py:362]: code path excited via PRO

File: devito/passes/clusters/buffering.py [hunk 5]
```diff
@@ -462,6 +424,85 @@ def __init__(self, function, dim, d, accessv, cache, options, sregistry):
             except TypeError:
                 self.buffer = cache[function] = Array(**kwargs)
 
+    def __init_multi_buffering__(self):
```
Comment (FabioLuporini) [inline devito/passes/clusters/buffering.py:427]: this is basically the only true new part

File: devito/passes/clusters/buffering.py [hunk 6]
```diff
@@ -462,6 +424,85 @@ def __init__(self, function, dim, d, accessv, cache, options, sregistry):
...
+        if self.is_read:
+            if is_integer(idx0) or isinstance(idx0, ModuloDimension):
+                # This is just for aesthetics of the generated code
+                self.index_mapper[idx1] = 0
+            else:
+                self.index_mapper[idx1] = idx1
+        else:
+            self.index_mapper[idx0] = idx1
+
+    def __init_firstlevel_buffering__(self, async_degree, sregistry):
```
Comment (FabioLuporini) [inline devito/passes/clusters/buffering.py:449]: already existing, simply refactored inside this method

File: devito/passes/clusters/asynchrony.py [hunk 7]
```diff
@@ -257,7 +258,7 @@ def _actions_from_update_memcpy(self, cluster, clusters, prefix, actions):
 
         # If fetching into e.g. `ub[sb1]` we'll need to prefetch into e.g. `ub[sb0]`
         tindex0 = e.lhs.indices[d]
-        if is_integer(tindex0):
+        if is_integer(tindex0) or isinstance(tindex0, IntDiv):
```
Comment (mloubout) [inline devito/passes/clusters/asynchrony.py:261]: Maybe add `or isintance(tindex0, Mod)` for case someone use % instead of IntDiv

Comment (FabioLuporini) [reply devito/passes/clusters/asynchrony.py:261]: for that you end up in the `else`, deliberately :) 

what you see here isn't stuff that the user has written, but rather by-product expressions of other passes such as buffering

**Pull request merged**